src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,1311 @@
+}
+
+LedgerStateHeader
+LedgerState::loadHeader()
+{
+    return mImpl->loadHeader(*this);
+}
+
+LedgerStateHeader
+LedgerState::Impl::loadHeader(LedgerState& self)
+{
+    checkNotSealed();
+    checkNoChild();
+    if (mActiveHeader)
+    {
+        throw std::runtime_error("LedgerStateHeader is active");
+    }
+
+    mActiveHeader = std::make_shared<HeaderImpl>(self, mHeader);
+    return LedgerStateHeader(mActiveHeader);
+}
+
+std::vector<LedgerStateEntry>
+LedgerState::loadOffersByAccountAndAsset(AccountID const& accountID,
+                                         Asset const& asset)
+{
+    return mImpl->loadOffersByAccountAndAsset(*this, accountID, asset);
+}
+
+std::vector<LedgerStateEntry>
+LedgerState::Impl::loadOffersByAccountAndAsset(LedgerState& self,
+                                               AccountID const& accountID,
+                                               Asset const& asset)
+{
+    checkNotSealed();
+    checkNoChild();
+
+    std::vector<LedgerStateEntry> res;
+    auto offers = getOffersByAccountAndAsset(accountID, asset);
+    for (auto const& kv : offers)
+    {
+        auto const& key = kv.first;
+        res.emplace_back(load(self, key));
+    }
+    return res;
+}
+
+ConstLedgerStateEntry
+LedgerState::loadWithoutRecord(LedgerKey const& key)
+{
+    return mImpl->loadWithoutRecord(*this, key);
+}
+
+ConstLedgerStateEntry
+LedgerState::Impl::loadWithoutRecord(LedgerState& self, LedgerKey const& key)
+{
+    checkNotSealed();
+    checkNoChild();
+    if (mActive.find(key) != mActive.end())
+    {
+        throw std::runtime_error("Key already loaded");
+    }
+
+    auto newest = getNewestVersion(self, key);
+    if (!newest.entry)
+    {
+        return {};
+    }
+
+    auto impl =
+        std::make_shared<ConstLedgerStateEntryImpl>(self, *newest.entry);
+    mActive.emplace(key, impl);
+    return ConstLedgerStateEntry(impl);
+}
+
+void
+LedgerState::rollback()
+{
+    mImpl->rollback(getIdentifier());
+    mImpl.reset();
+}
+
+void
+LedgerState::Impl::rollback(Identifier id)
+{
+    if (mChild)
+    {
+        mChild->rollback();
+    }
+
+    mActive.clear();
+    mActiveHeader.reset();
+
+    mParent.rollbackChild(id);
+}
+
+void
+LedgerState::rollbackChild(Identifier id)
+{
+    mImpl->rollbackChild();
+}
+
+void
+LedgerState::Impl::rollbackChild()
+{
+    mChild = nullptr;
+}
+
+void
+LedgerState::unsealHeader(std::function<void(LedgerHeader&)> f)
+{
+    mImpl->unsealHeader(*this, f);
+}
+
+void
+LedgerState::Impl::unsealHeader(
+    LedgerState& self, std::function<void(LedgerHeader&)> f)
+{
+    if (!mIsSealed)
+    {
+        throw std::runtime_error("LedgerState is not sealed");
+    }
+    if (mActiveHeader)
+    {
+        throw std::runtime_error("LedgerStateHeader is active");
+    }
+
+    mActiveHeader = std::make_shared<HeaderImpl>(self, mHeader);
+    LedgerStateHeader header(mActiveHeader);
+    f(header.current());
+}
+
+void
+LedgerState::Impl::updateLastModified()
```

Comment:
This name makes me uneasy.

This function seals current object and *maybe* updates last modified fields for each entry. What about renaming it to `seal`?

Reply:
I don't like the name `seal` because it doesn't indicate that it will update the actual data stored in the `LedgerState`. Updated to `sealAndMaybeUpdateLastModified` since it accurately describes what this function does.